crypto/src/merkle.rs

```diff
@@ -346,16 +364,81 @@ mod tests {
         let hash2 = Hash::prehashed([2; Hash::LENGTH]).typed();
         let hash3 = Hash::prehashed([3; Hash::LENGTH]).typed();
         let hash4 = Hash::prehashed([4; Hash::LENGTH]).typed();
+        let hash5 = Hash::prehashed([5; Hash::LENGTH]).typed();
         assert!(hash1 < hash2 && hash2 < hash3 && hash3 < hash4);
 
-        let tree = [hash1, hash2, hash4]
+        let tree = [hash1, hash2, hash4, hash5]
             .into_iter()
             .collect::<MerkleTree<()>>();
         let tree = tree.add(hash3);
         assert_eq!(tree.get_leaf(0), Some(hash1));
         assert_eq!(tree.get_leaf(1), Some(hash2));
         assert_eq!(tree.get_leaf(2), Some(hash3));
         assert_eq!(tree.get_leaf(3), Some(hash4));
-        assert_eq!(tree.get_leaf(4), None);
+        assert_eq!(tree.get_leaf(4), Some(hash5));
+        assert_eq!(tree.get_leaf(5), None);
+    }
+
+    impl<T> MerkleTree<T> {
+        fn size(&self) -> usize {
+            self.iter().count()
+        }
+
+        fn depth(&self) -> u32 {
+            self.size().trailing_ones()
+        }
```

Comment:
Is this always true? I can't get it...

Reply:
Indeed, the method assumes that the tree is constructed `from_iterator` and has a neat triangular shape. Its only usage is `geometry` test, which is rather a description of implementation than a test

Reply:
I think at least comment about that would be helpful then